rfcs/2021-09-01-8547-accept-metrics-in-datadog-agent-source.md

```diff
@@ -0,0 +1,168 @@
+metrics from a Vector perspective so they can be benefit from Vector capabilities.
+
+## Context
+- Vector is foreseen as a Datadog Agents aggregator, thus receiving metrics from Datadog Agents is a logical development
+- Vector has support to send metrics to Datadog, thus receiving metrics from Agent is a consistent feature to add
+
+## Cross cutting concerns
+
+Some known issues should be handled by the work described here, and should not be worked on alone: #7283, #8626 & #8493.
+
+## Scope
+
+### In scope
+
+- Implement a Datadog metrics endpoint in Vector, it will match the [metrics intake
+  API](https://docs.datadoghq.com/api/latest/metrics/) with additional route that the Agent uses
+- Include support for sketches that uses protobuf.
+- Ensure all Datadog metrics type are mapped to internal Vector metric type and that there is no loss of accuracy in a
+  pass through configuration.
+
+### Out of scope
+
+- Anything not related to metrics
+  - Processing API validation requests
+  - Processing other kind of payloads: traces, event, etc.
+
+## Pain
+
+- Vector cannot fully aggregate Datadog Agent traffic
+- Inconsistency between Datadog logs vs. metrics support in Vector
+
+## Proposal
+
+### User Experience
+
+- Vector will support receiving Datadog Metrics sent by the official Datadog Agent through a standard source
+- Metrics received will be fully supported inside Vector, all metric types will be supported
+- The following metrics flow: `n*(Datadog Agents) -> Vector -> Datadog` should just work
+- No foreseen backward compatibily issue (tags management may be a bit bothersome)
+- New configuration settings should be consistent with existing ones
+
+Regarding the Datadog Agent configuration, ideally it should be only a matter of configuring `dd_url:
+https://vector.mycompany.tld` to forward metrics to a Vector deployement. However configuring this will lead to some
+non-metric payloads to be sent to Vector, while it seems not an option to change the Datadog Agent behavior, those
+non-metric payload should still be forwarded to Datadog.
+
+The `dd_url` endpoint configuration has a [conditionnal
+behavior](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config.go#L1199-L1201) (also
+[here](https://github.com/DataDog/datadog-agent/blob/main/pkg/forwarder/forwarder_health.go#L131-L143)). I.e. if
+`dd_url` contains a known pattern (i.e. it has a suffix that matches a Datadog site) some extra hostname manipulation
+happens. But overal, this conditional can be ignore here and if we want unmodified Datadog Agent to use a `dd_url`
+pointing to a Vector deployement, the following route will have to be supported:
+- `/api/v1/validate` for API key validation
+- `/api/v1/check_run` for check submission
+- `/intake/` for events and metadata (possibly others)
+- `/support/flare/` for support flare
+- `/api/v1/series` & `/api/beta/sketches` for metrics submission
```

Comment:
I don't think that would be a problem. Http request in the `datadog_agent` source ends up in [`build_events`](https://github.com/timberio/vector/blob/master/src/sources/datadog/agent.rs#L149-L155) and the url path is know there so we can act in consequence. The one part that might be a bit bothersome would be to build a warp filter based on a variable path list. And finally the `DatadogAgentSource` would have to issue request to Datadog to relay payload we won't process (this would need few new settings to be introduced, at least the region/endpoint where to send those payload possibly an API key to use to do that).